### PR TITLE
better warning filter for assert_*

### PR DESCRIPTION
**Pull request by mathause**
In #4864 I added a a decorator for the `xarray.testing.assert_*` functions to ensure warnings that were to errors (`pytest.mark.filterwarnings("error")`) do not error in `assert_*` (see https://github.com/pydata/xarray/pull/4760#issuecomment-774101639). As a solution I added

https://github.com/pydata/xarray/blob/5470d933452d88deb17cc9294a164c4a03f55dec/xarray/testing.py#L32

However, this is sub-optimal because this now removes all `ignore` filters! As dask stuff only gets evaluated in `assert_*` filters like `warnings.filterwarnings("ignore", "Mean of empty slice")` don't work for dask arrays!

I thought of setting

```python
warnings.simplefilter("ignore")
```

but this could suppress warnings we want to keep.

So now I remove all `"error"` warning filters and keep the rest. Note that the original filters get restored after `with warnings.catch_warnings():`. ().

---

I am not sure I expressed myself very clearly... let me know and I can try again. @keewis you had a look at #4864 maybe you can review this PR as well?

### CODE REVIEW

File: xarray/testing.py [hunk 1]
```diff
@@ -29,7 +29,8 @@ def wrapper(*args, **kwargs):
         __tracebackhide__ = True
 
         with warnings.catch_warnings():
-            warnings.simplefilter("always")
+            # only remove filters that would "error"
+            warnings.filters = [f for f in warnings.filters if f[0] != "error"]
```
Comment (keewis) [inline xarray/testing.py:33]: can we rely on `warnings.filters`? As far as I can tell, that's semi-public at best: the library is very consistent on using `_` as a prefix for truly private attributes, but it also doesn't document `filters` or list it in `__all__`. However, the standard library doesn't change very often, so I guess it's not too risky...

Comment (mathause) [reply xarray/testing.py:33]: Yes, I am not sure either. But I found no "public" way to unset a _specific_ filter. I think even if it breaks at one point it should mainly affect downstream libraries and not users.

The alternatives are
- always warn
- never warn
- remove the decorator

I would prefer the "result" of what I propose here but I also understand if you deem it too dangerous. I had a quick look at the python issues but found nothing in this direction.

Comment (keewis) [reply xarray/testing.py:33]: `warnings` will only change on upgrade to a new version of `python`, which involves some manual work, anyways (*and* testing is usually blocked by `numba` for a few months).

I guess this means it should be safe to use? @pydata/xarray, what do you think?

**Comment by mathause** [comment]
Ok merging this. I think we got some to fix it if this ever causes trouble for a future version of python.

cc @headtr1ck

**Pull request merged**